core/src/main/java/amino/run/kernel/client/KernelClient.java

```diff
@@ -28,41 +34,180 @@
+            len = s.readInt();
+            for (int i = 0; i < len; i += STEP_SIZE_IN_BYTES) {
+                /* Received data is useless. Just overwrites to same byte array */
+                s.read(randomBytes);
+            }
+        }
+    }
+
     /** Stub for the OMS */
     private OMSServer oms;
-    /** List of hostnames matched to kernel server stubs */
-    private Hashtable<InetSocketAddress, KernelServer> servers;
+
+    /* Maximum number of continuous heartbeat miss allowed */
+    private static final int MAX_HEARTBEAT_MISS_TIMES = 3;
+
+    /* Class to hold remote kernel server info. This class is accessible only within this outer class(KernelClient) */
+    private final class KernelServerInfo {
+        private InetSocketAddress serverAddress; /* Host address of kernel server */
+        private KernelServer remoteRef; /* Remote reference to the kernel server */
+        /* Below reference counts are used to remove a server when it fails to send heartbeats for
+        MAX_HEARTBEAT_MISS_ALLOWED times continuously */
+        /* Global reference count is incremented whenever heartbeat is attempted */
+        private int refCount;
+        private int refCountOnHeartBeat; /* Reference count upon successful heartbeats */
+
+        /* Minimum samples to consider data rates as consistent. */
+        private static final int MIN_STABLE_DATA_RATE_TIMES = 10;
+        private int stableDataRateTimes; /* Number of continuous stable data transfer rates */
+        private RandomData data; /* Amount of random data sent to server in heartbeats */
+
+        private int metricPollPeriod = MIN_METRIC_POLL_PERIOD_MS; /* Poll period */
+        private NodeMetric metric; /* Node metric to server */
+        private ResettableTimer metricsTimer; /* Metric timer for the server */
+
+        private KernelServerInfo(
+                InetSocketAddress serverAddress, KernelServer remoteRef, final NodeMetric metric) {
+            this.serverAddress = serverAddress;
+            this.remoteRef = remoteRef;
+            this.metric = metric;
+            data = new RandomData();
+            metricsTimer =
+                    new ResettableTimer(
+                            new TimerTask() {
+                                public void run() {
+                                    measureMetrics(KernelServerInfo.this);
+                                }
+                            },
+                            metricPollPeriod);
+            metricsTimer.start();
+        }
+    }
+
+    /** Map of remote kernel server hostname to their kernel server info */
+    private ConcurrentHashMap<InetSocketAddress, KernelServerInfo> servers;
+
+    /* Map of remote kernel server hostname to their metrics. It is maintained independent from servers map, so that
+    this map can be serialized and sent to OMS as is */
+    private ConcurrentHashMap<InetSocketAddress, NodeMetric> metrics;
```

Comment:
nit: I wonder whether it perhaps makes sense to merge metrics and servers?  Fewer concurrentHashMaps, fewer locks, etc.  Just a thought.

Reply:
Yeah. Removed it now. In the initial code,  didn't use the custom serialization and deserialization for `amino.run.kernel.common.ServerInfo`. It is an argument in kernel server heartbeats to OMS i.e., in `sendHeartBeat(ServerInfo srvinfo)` And this `KernelClient.metrics` map reference was stored and accessed from `ServerInfo` too as it is while sending kernel server heartbeats to OMS. It was done to avoid maintaining another completely new map of `ConcurrentHashMap<InetSocketAddress, NodeMetric>` in `ServerInfo` again and also to avoid metrics data copied to it from `KernelClient.metrics` each time heartbeat is sent to OMS.

            But now, have realized that we don't need to send all of these node metrics to OMS in each heartbeat. Can send only the ones which were collected since last heartbeat to OMS. Even the node metrics relative to each other remote server is measured at different times independently. So sending the newly collected metrics since last heartbeat can reduce the heartbeat message size to OMS. So, have written custom serialization and deserialization for `ServerInfo`. After this change, we don't need explicit `KernelClient.metrics` map anymore. we can have `NodeMetric` as a field in `KernelClient.KernelServerInfo` itself and thus requires single map of `ConcurrentHashMap<InetSocketAddress, KernelClient.KernelServerInfo> servers`